## 1 Eliminacja wartości odstających

means =

 Columns 1 through 6:

   1.8679e-01   1.4839e-02   2.1045e-01   2.0882e-01   7.9658e+01   1.0604e+00

 Column 7:

   9.0846e-03

ans = min 0.125 for class 1 is at 642
ans = min 0 for class 2 is at 642
ans = min 0 for class 3 is at 642
ans = min 0 for class 4 is at 642
ans = min -5.4538e-09 for class 5 is at 393
ans = min -1.214e-07 for class 6 is at 25
ans = min -3.26424e-09 for class 7 is at 539
ans = max 5.12127 for class 1 is at 186
ans = max 25.8229 for class 2 is at 186
ans = max 382.139 for class 3 is at 186
ans = max 380.858 for class 4 is at 186
ans = max 145297 for class 5 is at 186
ans = max 1934.12 for class 6 is at 186
ans = max 16.5702 for class 7 is at 186
means =

 Columns 1 through 6:

   1.8404e-01   6.9109e-04   9.4984e-04   1.5044e-05  -7.5376e-10   2.9695e-07

 Column 7:

  -9.6408e-11

ans = min 0.148148 for class 1 is at 186min 642 for class 
ans = min 3.51422e-09 for class 2 is at 704
ans = min 7.53088e-11 for class 3 is at 816
ans = min 5.13337e-11 for

Usuwam ze zbioru uczącego próbki 642 (wartości skrajnie małe dla większości klas) oraz 186 (wartości skrajnie duże dla większości klas). Sensowność usunięcia tych dwóch próbek widoczna jest skali wykresów generowanych przez plot2features. Przed eliminacją dane tworzyły dwa zgrupowania – punktów odstających i reszty. Po usunięciu wartości odstających skala wykresów pozwala zauważyć to jak wyglądają różnice między klasami.


In [ ]:
for i=1:size(Combs)
    ttemp = train;
    testtemp = test;

    first_idx = Combs(i, 1);
    second_idx = Combs(i, 2);

    ttemp = ttemp(:, [1 first_idx second_idx]);
    testtemp = testtemp(:, [1 first_idx second_idx]);

    % to nie jest najros�dniejszy wyb�r; 4 i 6 na pewno trzeba zmieni�
    % tutaj jawnie tworz� struktur� z parametrami dla klasyfikatora Bayesa
    % (po prawdzie, to dla funkcji licz�cej g�sto�� prawdobie�stwa) z za�o�eniem,
    % �e cechy s� niezale�ne

    pdfindep_para = para_indep(ttemp);
    pdfmulti_para = para_multi(ttemp);
    pdfparzen_para = para_parzen(ttemp, 0.001);
    % w sprawozdaniu trzeba podawa� szeroko�� okna (nie liczymy tego parametru z danych!)

    % wyniki do punktu 3
    base_ercf = zeros(1,3);
    base_ercf(1) = mean(bayescls(testtemp(:,2:end), @pdf_indep, pdfindep_para) != testtemp(:,1));
    base_ercf(2) = mean(bayescls(testtemp(:,2:end), @pdf_multi, pdfmulti_para) != testtemp(:,1));
    base_ercf(3) = mean(bayescls(testtemp(:,2:end), @pdf_parzen, pdfparzen_para) != testtemp(:,1));
    disp(Combs(i, :))
    disp(base_ercf)
end

  ### 2 Błędy dla kombinacji
   
   |kombinacja | pdfindep_para | pdfmulti_para | pdfparzen_para|
   |-----------|---------------|---------------|---------------|
   |2   3      |0.17818        |0.17873        |   0.21765     |
   |2   4      |0.0263158      |0.0049342      |0.0241228      |
   |2   5      |0.075658       |0.059759       |0.270833       |
   |2   6      |0.13706        |0.10197        |0.27686        |
   |2   7      |0.102522       |0.097588       |0.276864       |
   |2   8      |0.22752        |0.22862        |0.27686        |
   |3   4      |0.021382       |0.020833       |0.016996       |
   |3   5      |0.13158        |0.12774        |0.23794        |
   |3   6      |0.15570        |0.14912        |0.31469        |
   |3   7      |0.16228        |0.15406        |0.31469        |
   |3   8      |0.28125        |0.28399        |0.31469        |
   |4   5      |0.15406        |0.14583        |0.23958        |
   |4   6      |0.15844        |0.15132        |0.26864        |
   |4   7      |0.23136        |0.21217        |0.26864        |
   |4   8      |0.21272        |0.21272        |0.26864        |
   |5   6      |0.37500        |0.31305        |0.47149        |
   |5   7      |0.32511        |0.26919        |0.47149        |
   |5   8      |0.38158        |0.37829        |0.47149        |
   |6   7      |0.32675        |0.28947        |0.54331        |
   |6   8      |0.40077        |0.40241        |0.57346        |
   |7   8      |0.36842        |0.36678        |0.54331        |
   
Najbardziej korzystna jest zatem kobinacja cech 3 i 4.

Dla apriori = 0.25

  |kombinacja | pdfindep_para | pdfmulti_para | pdfparzen_para|
  |-----------|---------------|---------------|---------------|
  | 3  4      | 0.021382      | 0.020833      |0.016996       |
  | 4  6 | 0.158443  | 0.151316 |0.268640| 
  | 3  4 |0.021382|0.020833|0.016996|

## 3 Klasyfikacja w oparciu o dwie wybrane cechy

|cechy|pdfindep|pdfmulti|pdfparzen|
|--------|
|3  4|0.021382|0.020833|0.016996|

używając etykiet klienta

|cechy|pdfindep|pdfmulti|pdfparzen|
|3  4|0.010417|0.009868|0.006031|

## 4 redukcja zbioru trenującego

Uśrednione wyniki dla klasyfikacji z użyciem zredukowanego zbioru trenującego dla pięciu prób 

|czesc|cechy|pdfindep|std|min|max|pdfmulti|std|min|max|pdfparzen|std|min|max|
|----------|
|0.100000|3  4|0.027083|0.007248|0.020833|0.036184|0.026754|0.007892|0.020285|0.037829|0.039474|0.003595|0.033443|0.042215|
|0.250000|3  4|0.021382|0.000671|0.020833|0.022478|0.020175|0.001419|0.018640|0.021930|0.029276|0.002345|0.026316|0.032346|
|0.500000|3  4|0.021930|0.000775|0.021382|0.023026|0.020724|0.001521|0.019189|0.023026|0.020614|0.000490|0.020285|0.021382|

używając etykiet klienta

|czesc|cechy|pdfindep|std|min|max|pdfmulti|std|min|max|pdfparzen|std|min|max|
|----------|
|0.100000|3  4|0.011732|0.001883|0.009868|0.014254|0.009868|0.002260|0.007127|0.012610|0.026096|0.002377|0.022478|0.021382|
|0.250000|3  4|0.010855|0.000601|0.009868|0.011513|0.009539|0.001320|0.007675|0.010965|0.017105|0.001954|0.014254|0.021382|
|0.500000|3  4|0.009978|0.000715|0.009320|0.010965|0.008991|0.000490|0.008224|0.009320|0.008991|0.000831|0.008224|0.021382|

## 5 różne szerokości okna Parzena

|parzen width|error|
|-------------|
|0.000100|0.014254|
|0.000500|0.016996|
|0.001000|0.020285|
|0.005000|0.044408|
|0.010000|0.043311|

używając etykiet klienta

|parzen width|error|
|-------------|
|0.000100|0.004386|
|0.000500|0.005482|
|0.001000|0.008772|
|0.005000|0.031250|
|0.010000|0.030154|

## 6 Dwukrotnie większe prawdopodobieństwo apriori dla maści czarnych

Uśrednione dla pięciu prób wyniki klasyfikacji uzyskane przy użyciu zredukownego zbioru trenującego 

|cechy|pdfindep|std|min|max|pdfmulti|std|min|max|pdfparzen|std|min|max|
|----------|
|3 4|0.017836|0.001425|0.016813|0.019737|0.016520|0.001602|0.014620|0.019006|0.017982|0.001223|0.016082|0.019006|

etykiety klienta:

|cechy|pdfindep|std|min|max|pdfmulti|std|min|max|pdfparzen|std|min|max|
|----------|
|3  4|0.009649|0.000801|0.008772|0.010965|0.008041|0.000731|0.007310|0.008772|0.010673|0.000400|0.010234|0.010965|

## Normalizacja danych, czy potrzebna?

Odchylenie standardowe dla cech 3 i 4: 9.20618518943410E-04   9.51290337161051E-04

Odchylenie standardowe w poszczególnych klasach dla 

|klasa|std cecha 3|std cecha 4|
|-----|
|1|0.000063|0.000186|
|2|0.000162|0.000003|
|3|0.000022|0.000085|
|4|0.000012|0.000098|
|5|0.000023|0.000021|
|6|0.000255|0.000002|
|7|0.000010|0.000112|
|8|0.000126|0.000009|

błędy klasyfikacji dla znormalizowanych cech 3 i 4

|cechy|pdfindep|pdfmulti|pdfparzen|
|--------|
|3  4|0.021382|0.020833|0.020285|

błędy klasyfikacji dla znormalizowanych cech 3 i 4 korzystając z etykiet klienta

|cechy|pdfindep|pdfmulti|pdfparzen|
|----|
|3  4|0.010417|0.009868|0.004386|



